packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -3256,6 +3243,111 @@ export class ContainerRuntime
 		}
 	}
 
+	/**
+	 * This helper is called during summary submission. If there are pending ops, it does the following:
+	 * 1. If beforeSummaryGeneration is true, it will wait for pending ops to be processed.
+	 * 2. If there are still pending ops, it will return a failed summarize result (IBaseSummarizeResult)
+	 * unless this is the final summarize attempt and SkipFailingIncorrectSummary option is set.
+	 * @param logger - The logger to be used for sending telemetry.
+	 * @param referenceSequenceNumber - The reference sequence number of the summary attempt.
+	 * @param minimumSequenceNumber - The minimum sequence number of the summary attempt.
+	 * @param finalAttempt - Whether this is the final summary attempt.
+	 * @param beforeSummaryGeneration - Whether this is called before summary generation or after.
+	 * @returns failed summarize result (IBaseSummarizeResult) if summary should be failed, undefined otherwise.
+	 */
+	private async shouldFailOnPendingOps(
+		logger: ITelemetryLoggerExt,
+		referenceSequenceNumber: number,
+		minimumSequenceNumber: number,
+		finalAttempt: boolean,
+		beforeSummaryGeneration: boolean,
+	): Promise<IBaseSummarizeResult | undefined> {
+		if (!this.hasPendingMessages()) {
+			return;
+		}
+
+		// If called before generating summary, wait for pending ops to be processed before proceeding.
+		if (beforeSummaryGeneration) {
```

Comment:
The wait that occurs here isn't obvious from shouldFailOnPendingOps

Reply:
It feels like this should be a separate function. I think duplicate this.hasPendingMessages() checks are better than this wait code being hidden, up to you. Feel free to resolve

Reply:
Yeah, good point. The waiting part is unexpected. I moved it to a separate function because `submitSummary` is already too big. Let me see what I can do.

Reply:
Moved the waiting into the calling code.